PWAWidgets/README.md

```diff
@@ -0,0 +1,975 @@
+
+1. Let <var>event</var> be a new ExtendableEvent.
+1. Run the following steps in parallel:
+   1. Set <var>event["data"]</var> to a new object.
+   1. Set <var>event["host"]</var> to the id of the Widget Host bound to <var>message</var>.
+   1. If <var>message</var> is a request to refresh all widgets
+      1. Set <var>event["action"]</var> to "WidgetResume".
+      1. Return <var>event</var>.
+   1. Else if <var>message</var> is a request to install a widget, set <var>event["action"]</var> to "WidgetInstall".
+   1. Else if <var>message</var> is a request to uninstall a widget, set <var>event["action"]</var> to "WidgetUninstall".
+   1. Else if <var>message</var> is a request to update a widget’s settings, set <var>event["action"]</var> to "WidgetSave".
+   1. Else set <var>event["action"]</var> to the user action bound to <var>message</var>.
+   1. Let <var>instanceId</var> be the id of the Widget Instance bound to <var>message</var>.
+   1. Set <var>event["instance"]</var> to <var>instanceId</var>.
+   1. Let <var>widget</var> be the result of running the algorithm specified in [getByInstance(instanceId)](#widgetsgetbyinstance) with <var>instanceId</var>.
+   1. Set <var>event["tag"]</var> to <var>widget["tag"]</var>.
+   1. If <var>message</var> includes bound data,
+      1. Set <var>event["data"]</var> to the data value bound to <var>message</var>.
+1. Return <var>event</var>
+
+
+### WidgetInstall
+
+When the User Agent receives a request to create a new instance of a widget, it will need to create a placeholder for the instance before triggering the WidgetClick event within the Service Worker.
+
+Required `WidgetEvent` data:
+
+* `host`
+* `instance`
+* `tag`
+
+The <b id="creating-a-placeholder-instance">steps for creating a placeholder instance</b> with WidgetEvent <var>event</var>:
+
+1. Let <var>tag</var> be <var>event["widget"]["tag"]</var>.
+1. Let <var>widget</var> be the result of running the algorithm specified in [getByTag(tag)](#widgetsgetbytag) with <var>tag</var>.
+1. If <var>widget</var> is undefined, exit.
+1. Let <var>payload</var> be an object.
+1. Set <var>payload["data"]</var> to an empty JSON object.
+1. Set <var>payload["settings"]</var> to the result of [creating a default `WidgetSettings` object](#creating-a-default-widgetsettings-object) with <var>widget</var>.
+1. Set <var>payload["definition"]</var> to the value of <var>widget["definition"]</var>.
+1. Set <var>payload</var> to the result of [injecting manifest members into a `WidgetPayload`](#injecting-manifest-members-into-a-payload) with <var>payload</var>.
+1. Let <var>instance</var> be the result of [creating an instance](#creating-a-widget-instance) with <var>event["widget"]["id"]</var>, <var>event["widget"]["host"]</var>, and <var>payload</var>.
+1. Append <var>instance</var> to <var>widget["instances"]</var>.
+
+<p id="install">Here is the flow for install:</p>
+
+![](media/install.gif)
```

Comment:
Alt text here?

Reply:
The text below fully describes the image, which is why I didn’t include it.